test/python/quantum_info/test_synthesis.py

```diff
@@ -1542,6 +1546,43 @@ def test_opt_a1a2(self, nqubits):
             ccirc.count_ops().get("cx"), (23 / 48) * 4**nqubits - (3 / 2) * 2**nqubits + 4 / 3
         )
 
+    def test_cs_decomp_error_raises(self):
+        """
+        Test that either an exception is raised or the decomposition succeeds.
+        """
+        base_gate_dim = 4
+        num_ctrls = 4
+        for i in range(3, 5):  # on my system i=3 induced exception but may vary
+            base_gate = random_unitary(base_gate_dim, seed=i * 2089).data
+            cmat = _compute_control_matrix(base_gate, num_ctrls)
+            try:
+                qc = self.qsd(cmat)
+            except QiskitError as err:
+                if "CS decomposition error" in repr(err):
+                    pass
+                else:
+                    self.fail("unexpected QiskitError")
+            else:
+                cqcop = Operator(qc)
+                self.assertTrue(cqcop == Operator(cmat))
+
+    def test_block_diagonal(self):
+        """
+        Test catching block diagonal input matrices.
+        """
+        num_qubits = 4
+        dim = 2**num_qubits
+        halfdim = dim // 2
+        u1 = scipy.stats.unitary_group.rvs(halfdim, random_state=422)
+        u2 = scipy.stats.unitary_group.rvs(halfdim, random_state=423)
+        zmat = np.zeros((halfdim, halfdim))
+        mat = np.block([[u1, zmat], [zmat, u2]])
+        circ = self.qsd(mat)
+        ccirc = transpile(circ, basis_gates=["u", "cx"], optimization_level=0)
+        self.assertTrue(Operator(mat) == Operator(ccirc))
+        print(ccirc.count_ops(), self._qsd_l2_a1a2_mod(num_qubits))
+        self.assertLess(ccirc.count_ops().get("cx"), self._qsd_l2_a1a2_mod(num_qubits))
```

Comment:
```suggestion
        self.assertEqual(Operator(mat), Operator(ccirc))
        self.assertLess(ccirc.count_ops().get("cx"), self._qsd_l2_a1a2_mod(num_qubits))
```

Reply:
Although that said, I don't understand why we want the last assertion here - don't we want the QSD to find the smaller number of CXs?

Reply:
This change to the algorithm, which looks at the form of the incoming matrix, will find smaller number of CXs than the stock algorithm for this special case.

Reply:
Ahhh, I see - `_qsd_l2_a1a2_mod` is the stock algorithm, but the full path via `transpile` should use the full form?

Reply:
I'm not sure I understand what you mean by full form? Previously, by default, any call to `qs_decomposition` will run with a cx count of `_qsd_l2_a1a2_mod`. This change will improve on that in special cases independent of whether opt_a1 or opt_a2 are set.